src/Aspire.Hosting/Dcp/DcpExecutor.cs

```diff
@@ -1019,80 +1072,110 @@ private static void SetInitialResourceState(IResource resource, IAnnotationHolde
-                    resourceLogger.LogError(ex, "Failed to create resource {ResourceName}", resource.Name);
-                    await _executorEvents.PublishAsync(new OnResourceFailedToStartContext(cancellationToken, resourceType, resource, DcpResourceName: null)).ConfigureAwait(false);
                 }
             }
-
-            var tasks = new List<Task>();
-            foreach (var group in executableResources.GroupBy(e => e.ModelResource))
+            catch (Exception ex)
             {
-                // Force this to be async so that blocking code does not stop other executables from being created.
-                tasks.Add(Task.Run(() => CreateResourceExecutablesAsyncCore(group.Key, group, cancellationToken), cancellationToken));
+                // The purpose of this catch block is to ensure that if an error processing the overall
+                // configuration of the executable resource files. This is different to the exception handling
+                // block above because at this stage of processing we don't necessarily have any replicas
+                // yet. For example if a dependency fails to start.
+                resourceLogger.LogError(ex, "Failed to create resource {ResourceName}", resource.Name);
+                await _executorEvents.PublishAsync(new OnResourceFailedToStartContext(cancellationToken, resourceType, resource, DcpResourceName: null)).ConfigureAwait(false);
             }
+        }
+
+        var tasks = new List<Task>();
+        foreach (var group in executables.GroupBy(e => e.ModelResource))
+        {
+            // Force this to be async so that blocking code does not stop other executables from being created.
+            tasks.Add(Task.Run(() => CreateResourceExecutablesAsyncCore(group.Key, group, cancellationToken), cancellationToken));
+        }
+
+        return Task.WhenAll(tasks).WaitAsync(cancellationToken);
+    }
+
+    private Task CreateContainerExecutablesAsync(IEnumerable<AppResource> containerExecAppResources, CancellationToken cancellationToken)
+        => CreateSnapshotableResourcesAsync(CreateContainerExecutableAsync, containerExecAppResources, cancellationToken);
+
+    private Task CreateExecutablesAsync(IEnumerable<AppResource> execAppResources, CancellationToken cancellationToken)
+        => CreateSnapshotableResourcesAsync(CreateExecutableAsync, execAppResources, cancellationToken);
+
+    private async Task CreateContainerExecutableAsync(AppResource er, ILogger resourceLogger, CancellationToken cancellationToken)
+    {
+        if (er.DcpResource is not ContainerExec containerExe)
+        {
+            throw new InvalidOperationException($"Expected an {nameof(ContainerExec)} resource, but got {er.DcpResource.Kind} instead");
+        }
+        var spec = containerExe.Spec;
 
-            return Task.WhenAll(tasks).WaitAsync(cancellationToken);
+        try
+        {
+            AspireEventSource.Instance.DcpContainerExecutablesCreateStart();
```

Comment:
I like what you did in terms of tracking actual time to create the Executable and ContainerExec resources, but this means these event pairs (or "activities" in EventSource parlance) are reported multiple times and track individual resource creation. So my suggestion would be to add a "resourceName" parameter to the event (for both events), so that we can track which resources took how much time to create.